chia/wallet/dao_wallet/dao_wallet.py

```diff
@@ -0,0 +1,2268 @@
+        unconfirmed_removals: Dict[bytes32, Coin] = await self.wallet_state_manager.unconfirmed_removals_for_wallet(
+            self.wallet_info.id
+        )
+        coins = await select_coins(
+            spendable_amount,
+            coin_selection_config,
+            spendable_coins,
+            unconfirmed_removals,
+            self.log,
+            uint128(amount),
+        )
+        assert sum(c.amount for c in coins) >= amount
+        return coins
+
+    async def get_pending_change_balance(self) -> uint64:
+        # No spendable or receivable value
+        return uint64(0)
+
+    async def get_unconfirmed_balance(self, record_list: Optional[Set[WalletCoinRecord]] = None) -> uint128:
+        unc_bal = await self.wallet_state_manager.get_unconfirmed_balance(self.id(), record_list)
+        return uint128(unc_bal)
+
+    # if asset_id == None: then we get normal XCH
+    async def get_balance_by_asset_type(self, asset_id: Optional[bytes32] = None) -> uint128:
+        puzhash = get_p2_singleton_puzhash(self.dao_info.treasury_id, asset_id=asset_id)
+        records = await self.wallet_state_manager.coin_store.get_coin_records_by_puzzle_hash(puzhash)
+        return uint128(sum([cr.coin.amount for cr in records if not cr.spent]))
+
+    # if asset_id == None: then we get normal XCH
+    async def select_coins_for_asset_type(self, amount: uint64, asset_id: Optional[bytes32] = None) -> List[Coin]:
+        puzhash = get_p2_singleton_puzhash(self.dao_info.treasury_id, asset_id=asset_id)
+        records = await self.wallet_state_manager.coin_store.get_coin_records_by_puzzle_hash(puzhash)
+        # TODO: smarter coin selection algorithm
+        total = 0
+        coins = []
+        for record in records:
+            if not record.spent:
+                total += record.coin.amount
+                coins.append(record.coin)
+                if total >= amount:
+                    break
+        if total < amount:  # pragma: no cover
+            raise ValueError(f"Not enough of asset {asset_id}: {total} < {amount}")
+        return coins
+
+    async def coin_added(self, coin: Coin, height: uint32, peer: WSChiaConnection) -> None:
+        """
+        Notification from wallet state manager that a coin has been received.
+        This can be either a treasury coin update or funds added to the treasury
+        """
+        self.log.info(f"DAOWallet.coin_added() called with the coin: {coin.name().hex()}:{coin}.")
+        wallet_node: Any = self.wallet_state_manager.wallet_node
+        peer = wallet_node.get_full_node_peer()
+        if peer is None:  # pragma: no cover
+            raise ValueError("Could not find any peers to request puzzle and solution from")
+        try:
+            # Get the parent coin spend
+            cs = (await wallet_node.get_coin_state([coin.parent_coin_info], peer, height))[0]
+            parent_spend = await fetch_coin_spend(cs.spent_height, cs.coin, peer)
```

Comment:
We have a utility function for this probably better to use it: https://github.com/Chia-Network/chia-blockchain/blob/d966f3f9e63aed52dbd73544164202a9f11ce3d2/chia/wallet/util/wallet_sync_utils.py#L428

Reply:
same as previous comment - it looks to me like we're already using the `fetch_coin_spend` utility?

Reply:
I don't know what you did but this is NOT how it was >:(